_posts/2018-07-20-totw-153.md

```diff
@@ -0,0 +1,211 @@
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  using namespace ::testing;
+  EXPECT_THAT(..., proto::Partially(...)); // ::testing::proto::Partially
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+```
+
+This using-directive has basically **introduced a namespace alias** `proto` in
+the global namespace, roughly equivalent to the following:
+
+<pre class="prettyprint lang-cpp bad-code">
+namespace proto = ::testing::proto;
+
+namespace totw {
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  EXPECT_THAT(..., proto::Partially(...)); // ::testing::proto::Partially
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+</pre>
+
+The test will keep compiling until a header defining namespace `::proto`,
+`::totw::proto`, or `::totw::example::proto` gets included transitively. At that
+point in time, `proto::Partially` becomes ambiguous, and the test stops
+compiling. This ties into the style guide's rules on namespace naming: avoid
+nested namespaces, and don't use common names for nested namespaces. (See [TotW
+130](http://abseil.io/tips/130) and the style rules on [Namespace
+names](https://google.github.io/styleguide/cppguide.html#Namespace_Names) for
+more on this topic.)
+
+One might think that it's safe to employ a using-directive for a closed
+namespace that has few symbols and guarantees that no more symbols will be ever
+added to it. (`std::placeholders`, which contains symbols `_1` ... `_9`, is an
+example of such a namespace.) However, even that isn't safe: it precludes any
+other namespace from introducing symbols with the same names. In this sense,
+using-directives defeat the modularity provided by namespaces.
+
+## Unqualified using-directives
+
+We've seen how one using-directive can go wrong. What happens if we have many of
+them, unqualified, in the same codebase?
+
+```c++
+namespace totw {
+namespace example {
+namespace {
+
+using namespace rpc;
+using namespace testing;
+
+TEST(MyTest, UsesUsingDirectives) {
+  Sequence seq;  // ::testing::Sequence
+  WallTimer timer;  // ::WallTimer
+  RPC rpc;  // ...is this ::rpc::RPC or ::RPC?
+  ...
+}
+
+}  // namespace
+}  // namespace example
+}  // namespace totw
+```
+
+What could possibly go wrong here? A lot, as it turns out:
+
+*   All the problems from our function-level example still exist, but two-fold:
+    once for namespace `::testing`, and once for namespace `::rpc`.
+*   If namespace `::rpc` and namespace `::testing` declare symbols with the same
+    name, this code won't compile if it does unqualified lookup on one of those
+    names. This is important, because it demonstrates a terrifying scaling
+    problem: since the full contents of each namespace is (generally speaking)
+    injected into the global namespace, **every new using-directive could add
+    quadratic risk of name collisions and build failures.**
+*   If a sub-namespace such as `::rpc::testing` is ever introduced, this code
```

Comment:
This problem stems from using `testing` rather than `::testing` and isn't related to using directives. If one were to use a using-declaration for `testing::Sequence`, or even type this name everywhere explicitly, this problem would still exist.